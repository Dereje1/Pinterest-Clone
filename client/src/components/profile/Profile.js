import React, { useEffect, useState, useLayoutEffect } from 'react';
import { useParams, useLocation, useHistory } from 'react-router-dom';
import { useSelector } from 'react-redux';
import Typography from '@mui/material/Typography';
import Avatar from '@mui/material/Avatar';
import Button from '@mui/material/Button';
import ButtonGroup from '@mui/material/ButtonGroup';
import Box from '@mui/material/Box';
import ImageBuild from '../imagebuild/Imagebuild';
import Loading from '../imagebuild/Loading';
import SignIn from '../signin/signin';
import RESTcall from '../../crud';
import { getProviderIcons } from '../common/common';
import error from '../mypins/error.png';

const providerIcons = getProviderIcons({ fontSize: 25 });

function Profile() {
  const [pinsOwned, setPinsOwned] = useState([]);
  const [pinsSaved, setPinsSaved] = useState([]);
  const [ready, setReady] = useState(false);
  const [displaySetting, setDisplaySetting] = useState('created');
  const [displayLogin, setDisplayLogin] = useState(false);
  const [retrievedUser, setRetrievedUser] = useState({ userId: '', service: 'twitter', displayName: '' });

  const { userInfo } = useParams();
  const { pathname } = useLocation();
  const history = useHistory();
  const loggedInUser = useSelector((state) => state.user);

  const getProfileData = async () => {
    try {
      setReady(false);
      const {
        createdPins, savedPins, user, redirect,
      } = await RESTcall({ address: `/api/userProfile/${userInfo}` });
      if (redirect) {
        history.push(redirect);
        return;
      }
      setPinsOwned(createdPins);
      setPinsSaved(savedPins);
      setRetrievedUser(user);
      setReady(true);
    } catch (err) {
      setPinsOwned([]);
      setPinsSaved([]);
      setReady(false);
    }
  };

  useEffect(() => {
    getProfileData();
    setDisplaySetting('created');
  }, [pathname]);

  useEffect(() => {
    const userObjectIsAvailable = Boolean(Object.keys(loggedInUser).length);
    if (userObjectIsAvailable && !loggedInUser.authenticated) {
      setReady(true);
      setDisplayLogin(true);
    }
  }, [loggedInUser]);

  useLayoutEffect(() => {
    document.body.style.overflowY = 'scroll';
  });

  const pins = displaySetting === 'created' ? pinsOwned : pinsSaved;
  if (!ready) return <div style={{ marginTop: 100 }}><Loading /></div>;

  if (displayLogin) {
    return (
      <SignIn
        removeSignin={() => {
          history.push('/');
        }}
      />
    );
  }

  return (
    <>
      <div style={{
        display: 'flex',
        flexDirection: 'column',
        justifyContent: 'space-between',
        alignItems: 'center',
        marginTop: 60,
        marginLeft: 10,
        marginRight: 10,
      }}
      >
        <Typography
          variant="h4"
          color="text.secondary"
        >
          PROFILE
        </Typography>
        <Avatar sx={{ bgcolor: providerIcons[retrievedUser.service].color, mt: 3 }}>
          {providerIcons[retrievedUser.service].icon}
        </Avatar>
        <Typography variant="h6" sx={{ mt: 3 }}>{retrievedUser.displayName}</Typography>
        <ButtonGroup variant="text" aria-label="text button group" sx={{ mt: 3 }}>
          <Button
            color={displaySetting === 'created' ? 'secondary' : 'primary'}
            onClick={() => setDisplaySetting('created')}
          >
            Pins created
          </Button>
          <Button
            color={displaySetting === 'saved' ? 'secondary' : 'primary'}
            onClick={() => setDisplaySetting('saved')}
          >
            Pins saved
          </Button>
        </ButtonGroup>
      </div>

      { pins.length ? (
        <ImageBuild
          pinImage
          deletePin={null}
          pinList={pins}
          ready={ready}
          user={loggedInUser}
        />
      ) : (
        <div style={{
          display: 'flex',
          flexDirection: 'column',
          justifyContent: 'space-between',
          alignItems: 'center',
        }}
        >
          <Box sx={{
            marginLeft: 'auto',
            marginRight: 'auto',
            marginTop: 2,
            padding: 2,
            textAlign: 'center',
          }}
          >
            <Typography
              variant="h6"
              color="text.secondary"
            >
              {`${retrievedUser.displayName} has not ${displaySetting} any pins`}
            </Typography>
          </Box>

          <img
            alt="no-pins-created"
            src={error}
            style={{ width: 200, height: 200 }}
          />
        </div>
      )}

    </>
  );
}
export default Profile;
