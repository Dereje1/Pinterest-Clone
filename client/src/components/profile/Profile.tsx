import React, { useEffect, useState, useLayoutEffect } from 'react';
import { useParams, useLocation, useHistory } from 'react-router-dom';
import { useSelector } from 'react-redux';
import Typography from '@mui/material/Typography';
import Avatar from '@mui/material/Avatar';
import Box from '@mui/material/Box';
import ImageBuild from '../imagebuild/Imagebuild';
import { Loading, getProviderIcons, UserPinsSelector } from '../common/common';
import SignIn from '../signin/signin';
import SearchUsers from './SearchUsers';
import RESTcall from '../../crud';
import { providerIconsType, userType } from '../../interfaces';
import error from '../mypins/error.png';

const providerIcons = getProviderIcons({ fontSize: 25 });

function Profile() {
  const [pinsOwned, setPinsOwned] = useState([]);
  const [pinsSaved, setPinsSaved] = useState([]);
  const [ready, setReady] = useState(false);
  const [displaySetting, setDisplaySetting] = useState('created');
  const [displayLogin, setDisplayLogin] = useState(false);
  const [retrievedUser, setRetrievedUser] = useState({ service: 'twitter', displayName: '' });

  const { userInfo }:{userInfo: string} = useParams();
  const { pathname } = useLocation();
  const history = useHistory();
  const loggedInUser = useSelector(({ user }:{user: userType}) => user);

  const getProfileData = async () => {
    try {
      setReady(false);
      const {
        createdPins, savedPins, user, redirect,
      } = await RESTcall({ address: `/api/userProfile/${userInfo}` });
      if (redirect) {
        history.push(redirect);
        return;
      }
      setPinsOwned(createdPins);
      setPinsSaved(savedPins);
      setRetrievedUser(user);
      setReady(true);
    } catch (err) {
      setPinsOwned([]);
      setPinsSaved([]);
      setReady(false);
    }
  };

  useEffect(() => {
    getProfileData();
    setDisplaySetting('created');
  }, [pathname]);

  useEffect(() => {
    const userObjectIsAvailable = Boolean(Object.keys(loggedInUser).length);
    if (userObjectIsAvailable && !loggedInUser.authenticated) {
      setReady(true);
      setDisplayLogin(true);
    }
  }, [loggedInUser]);

  useLayoutEffect(() => {
    document.body.style.overflowY = 'scroll';
  });

  const pins = displaySetting === 'created' ? pinsOwned : pinsSaved;
  if (!ready) return <Loading />;

  if (displayLogin) {
    return (
      <SignIn
        removeSignin={() => {
          history.push('/');
        }}
      />
    );
  }

  return (
    <>
      <div style={{
        display: 'flex',
        flexDirection: 'column',
        justifyContent: 'space-between',
        alignItems: 'center',
        marginTop: 60,
        marginLeft: 10,
        marginRight: 10,
      }}
      >
        <SearchUsers />
        <Typography
          variant="h4"
          color="text.secondary"
        >
          PROFILE
        </Typography>
        <Avatar sx={{
          bgcolor: providerIcons[retrievedUser.service as keyof providerIconsType].color, mt: 3,
        }}
        >
          {providerIcons[retrievedUser.service as keyof providerIconsType].icon}
        </Avatar>
        <Typography variant="h6" sx={{ mt: 3 }}>{retrievedUser.displayName}</Typography>
        <UserPinsSelector
          displaySetting={displaySetting}
          setDisplaySetting={(val: string) => setDisplaySetting(val)}
        />
      </div>

      { pins.length ? (
        <ImageBuild
          pinImage
          deletePin={null}
          pinList={pins}
          ready={ready}
          user={loggedInUser}
        />
      ) : (
        <div style={{
          display: 'flex',
          flexDirection: 'column',
          justifyContent: 'space-between',
          alignItems: 'center',
        }}
        >
          <Box sx={{
            marginLeft: 'auto',
            marginRight: 'auto',
            marginTop: 2,
            padding: 2,
            textAlign: 'center',
          }}
          >
            <Typography
              variant="h6"
              color="text.secondary"
            >
              {`${retrievedUser.displayName} has not ${displaySetting} any pins`}
            </Typography>
          </Box>

          <img
            alt="no-pins-created"
            src={error}
            style={{ width: 200, height: 200 }}
          />
        </div>
      )}

    </>
  );
}
export default Profile;
