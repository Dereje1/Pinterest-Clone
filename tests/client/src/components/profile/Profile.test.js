/**
 * @jest-environment jsdom
 */
/* eslint-disable import/order */
/* eslint-disable import/first */
import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Profile from '../../../../../client/src/components/profile/Profile';

// Mock router hooks
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'), // use actual for all non-hook parts
  useParams: () => ({
    userInfo: 'stubId-twitter-stubName',
  }),
  useLocation: () => ({ pathname: '/stub-path-name' }),
  useHistory: jest.fn(),
}));

// Mock redux hooks
jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'), // use actual for all non-hook parts
  useSelector: jest.fn((() => ({ authenticated: true }))),
}));

// Mock REST calls
jest.mock('../../../../../client/src/crud');


import { useSelector } from 'react-redux';
import { useHistory } from 'react-router-dom';
import RESTcall from '../../../../../client/src/crud';

describe('The profile page', () => {
  afterEach(() => {
    useSelector.mockClear();
    RESTcall.mockClear();
  });
  test('Will render for pins created by the user', async () => {
    const wrapper = shallow(<Profile />);
    await Promise.resolve();
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('Will render for pins saved by the user', async () => {
    const wrapper = shallow(<Profile />);
    await Promise.resolve();
    const savedButton = wrapper.find({ color: 'primary' });
    savedButton.props().onClick();
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('Will render signin for non-authenticated users', async () => {
    // Mock redux hooks
    useSelector.mockImplementationOnce(() => ({ authenticated: false }));
    const wrapper = shallow(<Profile />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('Will render no image sign if user has not pinned or created any', async () => {
    RESTcall.mockImplementationOnce(() => Promise.resolve({
      createdPins: [],
      savedPins: [],
      user: {
        userId: 'stubId', service: 'twitter', displayName: 'stubName',
      },
    }));
    const wrapper = shallow(<Profile />);
    await Promise.resolve();
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('Will change the display setting', async () => {
    const wrapper = shallow(<Profile />);
    await Promise.resolve();
    // owned button is selected by default
    let ownedButton = wrapper.find('ForwardRef(Button)').at(0);
    let savedButton = wrapper.find('ForwardRef(Button)').at(1);
    expect(ownedButton.props().color).toBe('secondary');
    expect(savedButton.props().color).toBe('primary');
    // select savedbutton
    savedButton.props().onClick();
    ownedButton = wrapper.find('ForwardRef(Button)').at(0);
    savedButton = wrapper.find('ForwardRef(Button)').at(1);
    expect(ownedButton.props().color).toBe('primary'); // currently selected
    expect(savedButton.props().color).toBe('secondary');
    // select owned button
    ownedButton.props().onClick();
    ownedButton = wrapper.find('ForwardRef(Button)').at(0);
    savedButton = wrapper.find('ForwardRef(Button)').at(1);
    expect(ownedButton.props().color).toBe('secondary'); // currently selected
    expect(savedButton.props().color).toBe('primary');
  });

  test('Will redirect to server response if query can not be handled', async () => {
    const push = jest.fn();
    useHistory.mockImplementation(() => ({ push }));
    RESTcall.mockImplementationOnce(() => Promise.resolve({ redirect: '/serverresponse' }));
    shallow(<Profile />);
    await Promise.resolve();
    expect(push).toHaveBeenCalledWith('/serverresponse');
  });

  test('Will redirect to root if user chooses not to sign in', async () => {
    // Mock redux hooks
    const push = jest.fn();
    useSelector.mockImplementationOnce(() => ({ authenticated: false }));
    useHistory.mockImplementation(() => ({ push }));
    const wrapper = shallow(<Profile />);
    const signin = wrapper.find('SignIn');
    signin.props().removeSignin();
    expect(push).toHaveBeenCalledWith('/');
  });

  test('Will not render loading if not ready or REST call is rejected', async () => {
    RESTcall.mockImplementationOnce(() => Promise.reject());
    const wrapper = shallow(<Profile />);
    await Promise.resolve();
    expect(wrapper.find('Loading').length).toBe(1);
  });
});
