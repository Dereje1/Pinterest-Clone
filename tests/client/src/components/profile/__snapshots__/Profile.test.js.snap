// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The profile page Will render for pins created by the user 1`] = `
<Fragment>
  <div
    style={
      {
        "alignItems": "center",
        "display": "flex",
        "flexDirection": "column",
        "justifyContent": "space-between",
        "marginLeft": 10,
        "marginRight": 10,
        "marginTop": 60,
      }
    }
  >
    <ForwardRef(Typography)
      color="text.secondary"
      variant="h4"
    >
      PROFILE
    </ForwardRef(Typography)>
    <ForwardRef(Avatar)
      sx={
        {
          "bgcolor": "#1DA1F2",
          "mt": 3,
        }
      }
    >
      <Memo(ForwardRef(TwitterIcon))
        style={
          {
            "fontSize": 25,
          }
        }
      />
    </ForwardRef(Avatar)>
    <ForwardRef(Typography)
      sx={
        {
          "mt": 3,
        }
      }
      variant="h6"
    >
      stubName
    </ForwardRef(Typography)>
    <UserPinsSelector
      displaySetting="created"
      setDisplaySetting={[Function]}
    />
  </div>
  <ImageBuild
    deletePin={null}
    pinImage={true}
    pinList={
      [
        {
          "_id": "3",
          "comments": [],
          "createdAt": "2020-12-25T00:00:00.000+00:00",
          "hasSaved": false,
          "imgDescription": "imgDescription id-3",
          "imgLink": "imgLink id-3",
          "owner": {
            "name": "owner id-3",
            "service": "github",
            "userId": "3",
          },
          "owns": true,
          "savedBy": [
            {
              "name": "savedBy - id-3",
              "service": "google",
              "userId": "3",
            },
            {
              "name": "savedBy - id-3",
              "service": "github",
              "userId": "3",
            },
          ],
          "tags": [],
        },
      ]
    }
    ready={true}
    user={
      {
        "authenticated": true,
      }
    }
  />
</Fragment>
`;

exports[`The profile page Will render for pins saved by the user 1`] = `
<Fragment>
  <div
    style={
      {
        "alignItems": "center",
        "display": "flex",
        "flexDirection": "column",
        "justifyContent": "space-between",
        "marginLeft": 10,
        "marginRight": 10,
        "marginTop": 60,
      }
    }
  >
    <ForwardRef(Typography)
      color="text.secondary"
      variant="h4"
    >
      PROFILE
    </ForwardRef(Typography)>
    <ForwardRef(Avatar)
      sx={
        {
          "bgcolor": "#1DA1F2",
          "mt": 3,
        }
      }
    >
      <Memo(ForwardRef(TwitterIcon))
        style={
          {
            "fontSize": 25,
          }
        }
      />
    </ForwardRef(Avatar)>
    <ForwardRef(Typography)
      sx={
        {
          "mt": 3,
        }
      }
      variant="h6"
    >
      stubName
    </ForwardRef(Typography)>
    <UserPinsSelector
      displaySetting="saved"
      setDisplaySetting={[Function]}
    />
  </div>
  <ImageBuild
    deletePin={null}
    pinImage={true}
    pinList={
      [
        {
          "_id": "2",
          "comments": [],
          "createdAt": "2020-12-25T00:00:00.000+00:00",
          "hasSaved": true,
          "imgDescription": "imgDescription id-2",
          "imgLink": "imgLink id-2",
          "owner": {
            "name": "owner id-2",
            "service": "twitter",
            "userId": "2",
          },
          "owns": false,
          "savedBy": [
            {
              "name": "savedBy - id-2",
              "service": "google",
              "userId": "2",
            },
            {
              "name": "savedBy - id-2git",
              "service": "github",
              "userId": "2",
            },
          ],
          "tags": [
            {
              "_id": "tag_id_1",
              "tag": "tag 1",
            },
          ],
        },
      ]
    }
    ready={true}
    user={
      {
        "authenticated": true,
      }
    }
  />
</Fragment>
`;

exports[`The profile page Will render no image sign if user has not pinned or created any 1`] = `
<Fragment>
  <div
    style={
      {
        "alignItems": "center",
        "display": "flex",
        "flexDirection": "column",
        "justifyContent": "space-between",
        "marginLeft": 10,
        "marginRight": 10,
        "marginTop": 60,
      }
    }
  >
    <ForwardRef(Typography)
      color="text.secondary"
      variant="h4"
    >
      PROFILE
    </ForwardRef(Typography)>
    <ForwardRef(Avatar)
      sx={
        {
          "bgcolor": "#1DA1F2",
          "mt": 3,
        }
      }
    >
      <Memo(ForwardRef(TwitterIcon))
        style={
          {
            "fontSize": 25,
          }
        }
      />
    </ForwardRef(Avatar)>
    <ForwardRef(Typography)
      sx={
        {
          "mt": 3,
        }
      }
      variant="h6"
    >
      stubName
    </ForwardRef(Typography)>
    <UserPinsSelector
      displaySetting="created"
      setDisplaySetting={[Function]}
    />
  </div>
  <div
    style={
      {
        "alignItems": "center",
        "display": "flex",
        "flexDirection": "column",
        "justifyContent": "space-between",
      }
    }
  >
    <ForwardRef(Box)
      sx={
        {
          "marginLeft": "auto",
          "marginRight": "auto",
          "marginTop": 2,
          "padding": 2,
          "textAlign": "center",
        }
      }
    >
      <ForwardRef(Typography)
        color="text.secondary"
        variant="h6"
      >
        stubName has not created any pins
      </ForwardRef(Typography)>
    </ForwardRef(Box)>
    <img
      alt="no-pins-created"
      src={
        {
          "createTransformer": [Function],
        }
      }
      style={
        {
          "height": 200,
          "width": 200,
        }
      }
    />
  </div>
</Fragment>
`;

exports[`The profile page Will render signin for non-authenticated users 1`] = `
<SignIn
  removeSignin={[Function]}
/>
`;
